### PR TITLE
Support to set loadBalancerKey

### PR DESCRIPTION
**Pull request by ileler**
Support to override CachingSpringLoadBalancerFactory to customize the FeignLoadBalancer instance.
use example:
```
@Configuration
public class FeignClientConfiguration {

    @Bean
    @Primary
    public Client feignClient(SpringClientFactory clientFactory, @Autowired(required = false) LoadBalancedRetryFactory retryFactory) {
        return new LoadBalancerFeignClient(new Client.Default(null, null),
                new CustomCachingSpringLoadBalancerFactory(clientFactory, retryFactory), clientFactory);
    }

    class CustomCachingSpringLoadBalancerFactory extends CachingSpringLoadBalancerFactory {

        private volatile Map<String, FeignLoadBalancer> cache = new ConcurrentReferenceHashMap<>();

        public CustomCachingSpringLoadBalancerFactory(SpringClientFactory factory) {
            super(factory);
        }

        public CustomCachingSpringLoadBalancerFactory(SpringClientFactory factory, LoadBalancedRetryFactory loadBalancedRetryPolicyFactory) {
            super(factory, loadBalancedRetryPolicyFactory);
        }

        @Override
        public FeignLoadBalancer create(String clientName) {
            FeignLoadBalancer client = this.cache.get(clientName);
            if(client != null) {
                return client;
            }
            IClientConfig config = this.factory.getClientConfig(clientName);
            ILoadBalancer lb = this.factory.getLoadBalancer(clientName);
            ServerIntrospector serverIntrospector = this.factory.getInstance(clientName, ServerIntrospector.class);
            client = loadBalancedRetryFactory != null ? new RetryableFeignLoadBalancer(lb, config, serverIntrospector,
                    loadBalancedRetryFactory) {
                protected void customizeLoadBalancerCommandBuilder(final FeignLoadBalancer.RibbonRequest request, final IClientConfig config,
                                                                   final LoadBalancerCommand.Builder<FeignLoadBalancer.RibbonResponse> builder) {
                    builder.withServerLocator(request.getRequest().headers().get(Constants.HEADER_CLIENT_IP));
                }
            } : new FeignLoadBalancer(lb, config, serverIntrospector) {
                protected void customizeLoadBalancerCommandBuilder(final FeignLoadBalancer.RibbonRequest request, final IClientConfig config,
                                                                   final LoadBalancerCommand.Builder<FeignLoadBalancer.RibbonResponse> builder) {
                    builder.withServerLocator(request.getRequest().headers().get(Constants.HEADER_CLIENT_IP));
                }
            };
            this.cache.put(clientName, client);
            return client;
        }

    }

}
```

### CODE REVIEW

**Comment by marcingrzejszczak** [comment]
Why do you want to do that? What problem are you trying to solve? You have written no tests.

**Comment by marcingrzejszczak** [comment]
What are those specific requirements?

File: spring-cloud-openfeign-core/src/main/java/org/springframework/cloud/openfeign/ribbon/CachingSpringLoadBalancerFactory.java [hunk 1]
```diff
@@ -36,8 +36,8 @@
  */
 public class CachingSpringLoadBalancerFactory {
 
-	private final SpringClientFactory factory;
-	private LoadBalancedRetryFactory loadBalancedRetryFactory = null;
+	protected final SpringClientFactory factory;
+	protected LoadBalancedRetryFactory loadBalancedRetryFactory = null;
```
Comment (adriancole) [inline spring-cloud-openfeign-core/src/main/java/org/springframework/cloud/openfeign/ribbon/CachingSpringLoadBalancerFactory.java]: this probably should have been final, looking at the code

Comment (adriancole) [reply spring-cloud-openfeign-core/src/main/java/org/springframework/cloud/openfeign/ribbon/CachingSpringLoadBalancerFactory.java]: strictly speaking you don't need to expose these. your subtype can override and shadow the fields, or make similarly named ones.

**Comment by adriancole** [comment]
please change your example to actually use the DebugRule you mentioned. Right now, the two examples are not connected.

Also, please first try a composition approach. See if you can use the existing code to form what you need vs inheritance. For example, see if you can forward calls to existing types somehow, without subtyping.

Anyway, if you can't please copy/paste the least code possible so it is easy to see the entire integration, including DebugRule. We might be able to identify another way to achieve the same goal, which doesn't require updating two projects, and exposing their constructors, to achieve.

**Comment by ileler** [comment]
@adriancole Hello, I submitted a test case. Can you see my needs from here? If this PR does not pass, my needs seem to have no other way to achieve it.My requirement is actually to set the `loadBalancerKey` by overriding `FeignLoadBalancer`. Then filter the server by `loadBalancerKey` in the `IRule` instance of the `ribbon`.

**Comment by ileler** [comment]
Is there anyone tracking this pr?

**Comment by ryanjbaxter** [comment]
@adriancole want to take another pass at this?

**Comment by ileler** [comment]
@ryanjbaxter Hey, adriancole seems to be very busy and has no time to look at this PR. Can you find other members to evaluate whether to pass this PR? We have to improve efficiency, because it is not good to have other users waiting for this small feature to be officially released!

**Comment by ileler** [comment]
Well, thank you for your reply. Sorry to bother you, but this PR is still pending, and the reviewer needs to reply to confirm the status. Thank you for your comments, then please ask other members to consider whether to merge.

**Comment by ryanjbaxter** [comment]
Not sure what you mean, if you have another PR please submit it and we will take a look.

**Pull request merged**